### ISSUE TITLE
Test data leaked into meta training

### ISSUE THREAD

**Issue by xiangdal** [dragen1860/MAML-Pytorch#3]
In the `forward()` function of `maml.py` (see [here](https://github.com/dragen1860/MAML-Pytorch/blob/master/maml.py#L405)), the meta optimizer steps are taken no matter the input data is for training or for testing. 
This results in the leakage of testing examples to the meta-learner which needs to be guarded by the `training` flag.

**Comment by dragen1860** [comment]
NO.
in meta-learning setting, meta-training includes training and testing, meta-test includes training and testing as well. The label space between meta-training and meta-testing is separated properly. For sub-stage, say training and testing in meta-train or meta-testing, the `leakage` is normal, not an error.

**Comment by xiangdal** [comment]
Thank you for your response!
I am aware that the sub-stage update is achieved through `fast_weights`. My question is, at the end of the `forward()` function, does `self.meta_optim.step()` update parameters of the base learner, i.e., the \theta in the paper?

**Comment by dragen1860** [comment]
Yes, the meta_optim.step() will update \theta parameters only. The updates of fast_weights will be ignored.

**Comment by dragen1860** [comment]
Oh, maybe u r right.
During the test, no need to optimize the \theta parameters.
U can try it and see any problems.

**Comment by xiangdal** [comment]
I have already tried and it turned out the optimization step is required to be left out in meta testing. Otherwise it would give overly optimistic results for be trained on meta test sets.

**Comment by dragen1860** [comment]
Ok, thx. Feel free to submit a PR.

**Comment by dragen1860** [comment]
Hi, all. Please git pull to get the latest version which solved test data leaking bugs.
